weaver/core/drivers/fabric-driver/server/satp.ts

```diff
@@ -0,0 +1,363 @@
+    performLockRequest2['chaincode-id'] = 'satpsimpleasset';
+
+    // Locker and recipient
+    const locker = performLockRequest2['locker'];
+    const recipient = performLockRequest2['recipient'];
+    const lockerWalletPath = performLockRequest2['lockerWalletPath'];
+    const recipientWalletPath = performLockRequest2['recipientWalletPath'];
+    let hashFn = performLockRequest2['hash_fn'];
+    let hashBase64 = performLockRequest2['hashBase64'];
+    const targetNetwork = performLockRequest2['target-network'];
+    const channel = performLockRequest2['channel'];
+    const chaincodeId = performLockRequest2['chaincode-id'];
+
+    // Hash
+    let hash: HashFunctions.Hash
+    if (hashFn == 'SHA512') {
+        hash = new HashFunctions.SHA512()
+    } else {
+        hash = new HashFunctions.SHA256()
+    }
+
+    if (hashBase64) {
+        hash.setSerializedHashBase64(hashBase64)
+    }
+    else {
+        logger.info(`No hash provided, using random preimage`)
+    }
+
+    // Timeout
+    var timeout = 0, timeout2 = 0;
+    const currTime = Math.floor(Date.now() / 1000);
+    if (performLockRequest2['timeout-epoch']) {
+        let duration = performLockRequest2['timeout-epoch'] - currTime
+        timeout = performLockRequest2['timeout-epoch']
+        timeout2 = performLockRequest2['timeout-epoch'] + duration
+    }
+    else if (performLockRequest2['timeout-duration']) {
+        timeout = currTime + performLockRequest2['timeout-duration']
+        timeout2 = currTime + 2 * performLockRequest2['timeout-duration']
+    }
+
+    let gateway: Gateway = await getNetworkGateway(networkName);
+    const network: Network = await gateway.getNetwork(channel);
+    const contract = network.getContract(chaincodeId);
+
+    const params = performLockRequest2['param'].split(':')
+    const lockerCert = Buffer.from(lockerWalletPath).toString('base64')
+    const recipientCert = Buffer.from(recipientWalletPath).toString('base64')
+
+    var funcToCall, asset
+
+    if (performLockRequest2['fungible']) {
+        funcToCall = SatpAssetManager.createFungibleHTLC
+        asset = 'Fungible Asset'
+    } else {
+        funcToCall = SatpAssetManager.createHTLC
+        asset = 'Asset'
+    }
+
+    console.info(`Asset Lock: Lock ${asset}:\n`);
+    try {
+        console.info(`Trying ${asset} Lock: ${params[0]}, ${params[1]} by ${locker} for ${recipient}`)
+        const res = await funcToCall(contract,
+            params[0],
+            params[1],
+            recipientCert,
+            hash,
+            timeout,
+            null)
+        if (!res.result) {
+            throw new Error()
+        }
+        console.info(`${asset} Locked with Contract Id: ${res.result}, preimage: ${res.hash.getPreimage()}, hashvalue: ${res.hash.getSerializedHashBase64()}`)
+        console.info('Asset has been locked successfully')
+
+    } catch (error) {
+        console.error(`Could not Lock ${asset} in ${targetNetwork}`)
+    }
+
+    await gateway.disconnect()
```

Comment:
gateway should be disconnected in case of error (use try-catch-finally)

Reply:
Good catch, @outSH !

@outsidethecode This logic needs to be fixed in several locations in the file. You can use the try-catch-finally pattern as mentioned above or you should at least do the following: (1) ensure that if there's an error in the catch block preceding the gateway disconnection, then make sure that the catch block itself throws an error and that the control does not then try to do subsequent actions as if no error occurred, and (2) ensure that if you throw an error in the catch block for the caller, add an `await gateway.disconnect()` before doing so, for proper garbage collection.

Reply:
Fixed.

Reply:
Sorry, my notifications were suppressed, so I didn't see this message and was then distracted by other tasks.

@outsidethecode What @outSH suggests is correct. Can you fix this please?

Reply:
Thanks @outSH and @VRamakrishna. I've just committed the change to sample-satp.ts because we decided to rename satp.ts to sample-satp.ts.

Reply:
Resolving the conversation as it looks like the change request has been satisfied (and we can't merge unless all conversations are marked as 'resolved')